MAINT: remove exec_command() from build_ext

tylerjereddy opened the pull request:
Replace a single usage of `exec_command()` with a standard library `subprocess` usage--eventual goal is to completely remove `exec_command()` in favor of the standard library, but working in small / manageable chunks.

I'm slightly concerned to see that codecov shows the affected code block here as not covered--perhaps this is because we don't run / aggregate coverage results form Appveyor, but we certainly could (*should I do that?*)--I suspect this may cause the coverage diff to fail here even if everything else is green.

Or that code block may genuinely not be covered by tests I suppose.

rgommers commented:
Do we have a good reason for this? I see no linked issue.

The problem with this stuff is that it's untestable, so should only be touched to fix issues as needed.

charris commented:
@rgommers It was suggested by @pv to avoid problems with using shells other than bash or sh, see #11824.

rgommers commented:
okay thanks, that makes sense now

eric-wieser commented on numpy/distutils/command/build_ext.py:
```diff
@@ -558,9 +558,10 @@ def _libs_with_msvc_and_fortran(self, fcompiler, c_libraries,
-                    dir = o
+                try:
+                    dir = subprocess.check_output(['cygpath', '-w', dir])
+                except subprocess.CalledProcessError:
```
Are you sure you want this and not `OsError`, which is what is thrown if the command does not exist?

eric-wieser replied:
(or perhaps both)

tylerjereddy replied:
And probably other things too:
```Python
import subprocess
ret = subprocess.check_output(['garbage', 'not', 'a', 'command'])
```

`FileNotFoundError: [Errno 2] No such file or directory: 'garbage': 'garbage'`

Would be much better if this was all well-tested, obviously.

eric-wieser replied:
`FileNotFoundError` is a subclass of `OSError`, and doesn't exist at all on python 2

charris commented:
And also #11822.

tylerjereddy commented:
I added in the requested additional exception class.

I share the concerns re: testability -- is it worth trying to mock up some unit tests that try to mimic the intended behavior? That mocking stuff tends not to be pretty, though perhaps better than nothing for flushing through the codepath.

charris commented:
The pending pytest fix #11896 should fix the failing tests here.

charris commented:
For testing, I think we can just ask folks who had problems if things work for them.

tylerjereddy commented:
Rebased / force pushed.

I note that `exec_command` also does output / string processing after the system call is made, and while it appears to be safe to ignore that here, in the other cases I've worked on so far (PRs incoming) it was sometimes necessary to reconstitute some of that downstream processing after `subprocess.check_call()`.

tylerjereddy committed:
MAINT: remove exec_command from build_ext

* the single usage of exec_command in
distutils build_ext module has been
replaced with a standard library usage
of subprocess

tylerjereddy commented:
Rebased & revised to match https://github.com/numpy/numpy/pull/11900

eric-wieser commented:
Ignoring coverage failures, since this is distutils
